polkadot/xcm/xcm-executor/src/traits/asset_transfer.rs

```diff
@@ -0,0 +1,88 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::traits::TransactAsset;
+use frame_support::traits::ContainsPair;
+use scale_info::TypeInfo;
+use sp_runtime::codec::{Decode, Encode};
+use xcm::prelude::*;
+
+/// Errors related to determining asset transfer support.
+#[derive(Copy, Clone, Encode, Decode, Eq, PartialEq, Debug, TypeInfo)]
+pub enum Error {
+	/// Invalid non-concrete asset.
+	NotConcrete,
+	/// Reserve chain could not be determined for assets.
+	UnknownReserve,
+}
+
+/// Specify which type of asset transfer is required for a particular `(asset, dest)` combination.
+#[derive(Copy, Clone, PartialEq, Debug)]
+pub enum TransferType {
+	/// should teleport `asset` to `dest`
+	Teleport,
+	/// should reserve-transfer `asset` to `dest`, using local chain as reserve
+	LocalReserve,
+	/// should reserve-transfer `asset` to `dest`, using `dest` as reserve
+	DestinationReserve,
+	/// should reserve-transfer `asset` to `dest`, using remote chain `MultiLocation` as reserve
+	RemoteReserve(MultiLocation),
+}
+
+/// A trait for identifying asset transfer type based on `IsTeleporter` and `IsReserve`
+/// configurations.
+pub trait AssetTransferSupport {
```

Comment:
This also exposes
```
	/// How to withdraw and deposit an asset.
	type AssetTransactor: TransactAsset;
```
so it's more than just determining asset transfer type/strategy..

Reply:
I too have reservations about the naming, but aside from that, what I'm more interested in is how this trait composes with the `Config` trait. It would seem to me that it is just reassigning the 3 associated types in this trait. Given that's the case, there are a couple of things that we can do to eliminate redundancy between these 2 traits:

1. Remove the 3 associated type items in `Config`, and make `AssetTransferSupport` a supertrait of `Config`, i.e. implementors of `Config` MUST also implement `AssetTransferSupport`;
2. Remove the 3 associated type items here in `AssetTransferSupport`, and make `Config` a supertrait of `AssetTransferSupport`

I'd probably prefer (1) as the executor has to make use of `AssetTransactor`, `IsReserve` and `IsTeleporter` in its own processing; furthermore, this would also allow for a better syntax in the XCM pallet when trying to access `IsReserve` or `IsTeleporter`.

Reply:
It doesn't work that way - `xcm_executor::config::Config` is not actually exposed to pallets, but traits implemented on the executor itself are exposed (such as `ExecuteXcm`) - making `AssetTransferSupport` a supertrait of Config or even the other way around doesn't then help us in using it like:
```
type XcmExecutor: ExecuteXcm<<Self as Config>::RuntimeCall> + <need-some-custom-trait-here-anyway-cannot-be-config>;
```

More so, looks like `xcm_executor::config::Config` is _the general_ `XcmConfig` configured for runtimes - and it's pretty nice that it's flat and all XCM configurations go in a single place, splitting its types to subtraits just makes it harder for builders IMO (also is a breaking change at the code level).

Anyway I made a prototype of your suggestion at https://github.com/acatangiu/polkadot-sdk/commit/f40c21081a46f19f42b6d7f885648a24c9315035 - have a look - I am leaning towards keeping it as is. We duplicate a couple types in the implementation underbelly, but at the benefit of adding this enhancement transparently (no API breaks).

Also renamed the trait to `XcmAssetTransfers` hope it sounds better :)